homestar-runtime/src/event_handler/swarm_event.rs

```diff
@@ -187,26 +197,96 @@ async fn handle_swarm_event<THandlerErr: fmt::Debug + Send, DB: Database>(
                     registrations,
                     cookie,
                 } => {
-                    // save cookie for later (when we are hungry for snacks again. yummy.)
                     if cookie.namespace() == Some(&Namespace::from_static(RENDEZVOUS_NAMESPACE)) {
+                        debug!(
+                            peer_id = rendezvous_node.to_string(),
+                            "received discovery from rendezvous server"
+                        );
+
+                        // Store cookie
                         event_handler
                             .rendezvous_cookies
                             .insert(rendezvous_node, cookie);
 
-                        // dial discovered peers
-                        for registration in registrations {
-                            // TODO: do anything with ttl here?
-                            let opts = DialOpts::peer_id(registration.record.peer_id())
-                                .addresses(registration.record.addresses().to_vec())
-                                .condition(libp2p::swarm::dial_opts::PeerCondition::Disconnected)
-                                .build();
-                            // TODO: we might be dialing too many peers here. Add settings to configure when we stop dialing new peers
-                            if let Err(err) = event_handler.swarm.dial(opts) {
-                                warn!(peer_id=registration.record.peer_id().to_string(), err=?err, "failed to dial peer discovered through rendezvous")
+                        let connected_peers_count = event_handler.connected_peers.len();
+
+                        // Skip dialing peers if at connected peers limit
+                        if connected_peers_count >= event_handler.connected_peers_limit as usize {
+                            warn!("peers discovered through rendezvous not dialed because max connected peers limit reached");
+                            return;
+                        }
+
+                        // Filter out already connected peers
+                        let new_registrations: Vec<&Registration> = registrations
+                            .iter()
+                            .filter(|registration| {
+                                !event_handler
+                                    .connected_peers
+                                    .contains_key(&registration.record.peer_id())
+                            })
+                            .collect();
+
+                        // Dial newly discovered peers
+                        for (index, registration) in new_registrations.iter().enumerate() {
+                            let our_registration = &registration.record.peer_id()
```

Comment:
minor: our feels weird naming-wise. `self_registration`?

Reply:
Yup that is better. Renamed it. 👍